File: server-go/internal/jobs/marketsched/decision_test.go
package marketsched

import (
	"testing"
	"time"
)

// 2026-07-17 是周五,2026-07-18 周六,2026-07-19 周日(上海时区)。
func utc(hour, minute int) time.Time {
	return time.Date(2026, 7, 17, hour, minute, 30, 0, time.UTC)
}

func shanghai(hour, minute int) time.Time {
	return utc(hour-8, minute)
}

func onConfig() Config {
	return Config{ScheduleEnabled: true, LastIntervalMinutes: 60, SettlementEnabled: true}
}

func TestDecideScheduleDisabled(t *testing.T) {
	cfg := onConfig()
	cfg.ScheduleEnabled = false
	// 09:00 上海本是整点槽且在日盘时段,关总开关则一律不跑
	decision := Decide(cfg, shanghai(9, 0), State{})
	if decision.RunLasts || decision.RunSettlements {
		t.Fatalf("disabled decide = %#v", decision)
	}
	// 结算槽也不跑
	decision = Decide(cfg, shanghai(15, 30), State{})
	if decision.RunLasts || decision.RunSettlements {
		t.Fatalf("disabled settlement decide = %#v", decision)
	}
}

func TestDecideLastsSlots(t *testing.T) {
	cases := []struct {
		name string
		cfg  Config
		now  time.Time
		prev State
		want bool
	}{
		{"首次运行整点到槽", onConfig(), shanghai(9, 0), State{}, true},
		{"槽内第1分钟容忍漂移", onConfig(), shanghai(9, 1), State{}, true},
		{"槽内第2分钟不触发", onConfig(), shanghai(9, 2), State{}, false},
		{"非槽位不触发", onConfig(), shanghai(9, 30), State{}, false},
		{"同槽去重", onConfig(), shanghai(9, 0), State{Lasts: SlotKey{Date: "2026-07-17", Slot: 540}}, false},
		{"下一槽再触发", onConfig(), shanghai(10, 0), State{Lasts: SlotKey{Date: "2026-07-17", Slot: 540}}, true},
		{"日盘起点前不触发", onConfig(), shanghai(8, 0), State{}, false},
		{"日盘末段 15:00 触发", onConfig(), shanghai(15, 0), State{}, true},
		{"日盘结束后不触发", onConfig(), shanghai(16, 0), State{}, false},
		{"夜盘 21:00 触发", onConfig(), shanghai(21, 0), State{}, true},
		{"夜盘跨零点触发", onConfig(), shanghai(0, 0), State{}, true},
		{"夜盘末尾 02:30 非槽位不触发", onConfig(), shanghai(2, 30), State{}, false},
		{"间隔30分半点槽触发", Config{ScheduleEnabled: true, LastIntervalMinutes: 30, SettlementEnabled: true}, shanghai(9, 30), State{}, true},
		{"间隔120分两小时槽触发", Config{ScheduleEnabled: true, LastIntervalMinutes: 120, SettlementEnabled: true}, shanghai(10, 0), State{}, true},
		{"间隔120分奇数点不触发", Config{ScheduleEnabled: true, LastIntervalMinutes: 120, SettlementEnabled: true}, shanghai(9, 0), State{}, false},
		{"非法间隔按60分", Config{ScheduleEnabled: true, LastIntervalMinutes: 45, SettlementEnabled: true}, shanghai(10, 0), State{}, true},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			decision := Decide(tc.cfg, tc.now, tc.prev)
			if decision.RunLasts != tc.want {
				t.Fatalf("RunLasts = %v, want %v (decide=%#v)", decision.RunLasts, tc.want, decision)
			}
		})
	}
}

func TestDecideLastsStateAdvance(t *testing.T) {
	decision := Decide(onConfig(), shanghai(9, 0), State{})
	if !decision.RunLasts {
		t.Fatal("expected lasts run")
	}
	want := SlotKey{Date: "2026-07-17", Slot: 540}
	if decision.Next.Lasts != want {
		t.Fatalf("next lasts key = %#v, want %#v", decision.Next.Lasts, want)
	}
}

func TestDecideSettlementSlots(t *testing.T) {
	slots := []struct {
		hour, minute, wantSlot int
	}{
		{15, 30, 930}, {16, 0, 960}, {16, 30, 990}, {17, 0, 1020},
	}
	for _, slot := range slots {
		decision := Decide(onConfig(), shanghai(slot.hour, slot.minute), State{})
		if !decision.RunSettlements {
			t.Fatalf("%02d:%02d should trigger settlement", slot.hour, slot.minute)
		}
		want := SlotKey{Date: "2026-07-17", Slot: slot.wantSlot}
		if decision.Next.Settlement != want {
			t.Fatalf("next settlement key = %#v, want %#v", decision.Next.Settlement, want)
		}
	}
}

func TestDecideSettlementBoundaries(t *testing.T) {
	cases := []struct {
		name string
		cfg  Config
		now  time.Time
		prev State
		want bool
	}{
		{"槽后1分钟容忍漂移", onConfig(), shanghai(15, 31), State{}, true},
		{"槽后2分钟不触发", onConfig(), shanghai(15, 32), State{}, false},
		{"槽前不触发", onConfig(), shanghai(15, 29), State{}, false},
		{"同槽去重", onConfig(), shanghai(15, 30), State{Settlement: SlotKey{Date: "2026-07-17", Slot: 930}}, false},
		{"下一尝试槽重试", onConfig(), shanghai(16, 0), State{Settlement: SlotKey{Date: "2026-07-17", Slot: 930}}, true},
		{"结算开关关不触发", Config{ScheduleEnabled: true, LastIntervalMinutes: 60, SettlementEnabled: false}, shanghai(15, 30), State{}, false},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			decision := Decide(tc.cfg, tc.now, tc.prev)
			if decision.RunSettlements != tc.want {
				t.Fatalf("RunSettlements = %v, want %v (decide=%#v)", decision.RunSettlements, tc.want, decision)
			}
		})
	}
}

func TestDecideSettlementWeekendSkipped(t *testing.T) {
	// 2026-07-18 周六 15:30 上海 = 07:30 UTC
	saturday := time.Date(2026, 7, 18, 7, 30, 0, 0, time.UTC)
	if decision := Decide(onConfig(), saturday, State{}); decision.RunSettlements {
		t.Fatal("周六不应触发结算补拉")
	}
	sunday := time.Date(2026, 7, 19, 7, 30, 0, 0, time.UTC)
	if decision := Decide(onConfig(), sunday, State{}); decision.RunSettlements {
		t.Fatal("周日不应触发结算补拉")
	}
}

func TestDecideFirstRunOutsideSlots(t *testing.T) {
	// 首次运行(零状态)但不在任何槽位:不补跑,等下一槽
	decision := Decide(onConfig(), shanghai(9, 45), State{})
	if decision.RunLasts || decision.RunSettlements {
		t.Fatalf("decide = %#v", decision)
	}
}

File: server/src/jobs/marketsched/decision.test.ts
import { describe, expect, test } from 'bun:test'
import { decide, emptyScheduleState, type MarketScheduleConfig } from './decision.ts'

/** 2026-07-17 周五；shanghai(h,m) → 对应 UTC */
function shanghai(hour: number, minute: number): Date {
  return new Date(Date.UTC(2026, 6, 17, hour - 8, minute, 30))
}

function onConfig(over: Partial<MarketScheduleConfig> = {}): MarketScheduleConfig {
  return {
    scheduleEnabled: true,
    lastIntervalMinutes: 60,
    settlementEnabled: true,
    ...over,
  }
}

describe('marketsched decide', () => {
  test('定时总开关关不跑', () => {
    const cfg = onConfig({ scheduleEnabled: false })
    let d = decide(cfg, shanghai(9, 0), emptyScheduleState())
    expect(d.runLasts).toBe(false)
    expect(d.runSettlements).toBe(false)
    d = decide(cfg, shanghai(15, 30), emptyScheduleState())
    expect(d.runLasts).toBe(false)
    expect(d.runSettlements).toBe(false)
  })

  test('最新价槽位触发与去重', () => {
    const cases: Array<{
      name: string
      cfg?: Partial<MarketScheduleConfig>
      now: Date
      prev: ReturnType<typeof emptyScheduleState>
      want: boolean
    }> = [
      { name: '首次运行整点到槽', now: shanghai(9, 0), prev: emptyScheduleState(), want: true },
      { name: '槽内第1分钟容忍漂移', now: shanghai(9, 1), prev: emptyScheduleState(), want: true },
      { name: '槽内第2分钟不触发', now: shanghai(9, 2), prev: emptyScheduleState(), want: false },
      { name: '非槽位不触发', now: shanghai(9, 30), prev: emptyScheduleState(), want: false },
      {
        name: '同槽去重',
        now: shanghai(9, 0),
        prev: { lasts: { date: '2026-07-17', slot: 540 }, settlement: null },
        want: false,
      },
      {
        name: '下一槽再触发',
        now: shanghai(10, 0),
        prev: { lasts: { date: '2026-07-17', slot: 540 }, settlement: null },
        want: true,
      },
      { name: '日盘起点前不触发', now: shanghai(8, 0), prev: emptyScheduleState(), want: false },
      { name: '日盘末段 15:00 触发', now: shanghai(15, 0), prev: emptyScheduleState(), want: true },
      { name: '日盘结束后不触发', now: shanghai(16, 0), prev: emptyScheduleState(), want: false },
      { name: '夜盘 21:00 触发', now: shanghai(21, 0), prev: emptyScheduleState(), want: true },
      { name: '夜盘跨零点触发', now: shanghai(0, 0), prev: emptyScheduleState(), want: true },
      {
        name: '夜盘末尾 02:30 非槽位不触发',
        now: shanghai(2, 30),
        prev: emptyScheduleState(),
        want: false,
      },
      {
        name: '间隔30分半点槽触发',
        cfg: { lastIntervalMinutes: 30 },
        now: shanghai(9, 30),
        prev: emptyScheduleState(),
        want: true,
      },
      {
        name: '间隔120分两小时槽触发',
        cfg: { lastIntervalMinutes: 120 },
        now: shanghai(10, 0),
        prev: emptyScheduleState(),
        want: true,
      },
      {
        name: '间隔120分奇数点不触发',
        cfg: { lastIntervalMinutes: 120 },
        now: shanghai(9, 0),
        prev: emptyScheduleState(),
        want: false,
      },
      {
        name: '非法间隔按60分',
        cfg: { lastIntervalMinutes: 45 },
        now: shanghai(10, 0),
        prev: emptyScheduleState(),
        want: true,
      },
    ]
    for (const tc of cases) {
      const d = decide(onConfig(tc.cfg), tc.now, tc.prev)
      expect(d.runLasts, tc.name).toBe(tc.want)
    }
  })

  test('最新价状态前进', () => {
    const d = decide(onConfig(), shanghai(9, 0), emptyScheduleState())
    expect(d.runLasts).toBe(true)
    expect(d.next.lasts).toEqual({ date: '2026-07-17', slot: 540 })
  })

  test('结算槽位触发', () => {
    for (const [hour, minute, wantSlot] of [
      [15, 30, 930],
      [16, 0, 960],
      [16, 30, 990],
      [17, 0, 1020],
    ] as const) {
      const d = decide(onConfig(), shanghai(hour, minute), emptyScheduleState())
      expect(d.runSettlements).toBe(true)
      expect(d.next.settlement).toEqual({ date: '2026-07-17', slot: wantSlot })
    }
  })

  test('结算边界与开关', () => {
    const cases = [
      {
        name: '槽后1分钟容忍漂移',
        now: shanghai(15, 31),
        prev: emptyScheduleState(),
        cfg: {},
        want: true,
      },
      {
        name: '槽后2分钟不触发',
        now: shanghai(15, 32),
        prev: emptyScheduleState(),
        cfg: {},
        want: false,
      },
      {
        name: '槽前不触发',
        now: shanghai(15, 29),
        prev: emptyScheduleState(),
        cfg: {},
        want: false,
      },
      {
        name: '同槽去重',
        now: shanghai(15, 30),
        prev: { lasts: null, settlement: { date: '2026-07-17', slot: 930 } },
        cfg: {},
        want: false,
      },
      {
        name: '下一尝试槽重试',
        now: shanghai(16, 0),
        prev: { lasts: null, settlement: { date: '2026-07-17', slot: 930 } },
        cfg: {},
        want: true,
      },
      {
        name: '结算开关关不触发',
        now: shanghai(15, 30),
        prev: emptyScheduleState(),
        cfg: { settlementEnabled: false },
        want: false,
      },
    ] as const
    for (const tc of cases) {
      const d = decide(onConfig(tc.cfg), tc.now, tc.prev)
      expect(d.runSettlements, tc.name).toBe(tc.want)
    }
  })

  test('周末跳过结算', () => {
    // 2026-07-18 周六 15:30 上海 = 07:30 UTC
    const saturday = new Date(Date.UTC(2026, 6, 18, 7, 30, 0))
    expect(decide(onConfig(), saturday, emptyScheduleState()).runSettlements).toBe(false)
    const sunday = new Date(Date.UTC(2026, 6, 19, 7, 30, 0))
    expect(decide(onConfig(), sunday, emptyScheduleState()).runSettlements).toBe(false)
  })

  test('首次运行不在槽位不补跑', () => {
    const d = decide(onConfig(), shanghai(9, 45), emptyScheduleState())
    expect(d.runLasts).toBe(false)
    expect(d.runSettlements).toBe(false)
  })
})
